onnxruntime/core/providers/cpu/llm/attention.cc

```diff
@@ -0,0 +1,749 @@
+            } else {
+              v = present + ki * present_chunk_length;
+            }
+          } else if (nullptr != present_value) {
+            if (past_present_share_buffer) {
+              v = present_value + cache_chunk_length * ki;
+              if (head_i < kv_num_heads) {
+                memcpy(const_cast<T*>(v) + past_chunk_length, V + v_head_size * ki, v_head_size * sizeof(T));
+              }
+            } else {
+              if (head_i < kv_num_heads) {
+                v = ConcatStateChunk(past_value, v, present_value, past_chunk_length, present_chunk_length, ki);
+              } else {
+                v = present_value + ki * present_chunk_length;
+              }
+            }
+          }
+
+          if (transpose_output) {
+            T* current_tmp_data = reinterpret_cast<T*>(tmp_buffer) + q_input_chunk_length * i;
+            ptrdiff_t attention_probs_offset = SafeInt<ptrdiff_t>(sequence_length) * total_sequence_length * i;
+            math::MatMul<T>(sequence_length, v_head_size, total_sequence_length,
+                            attention_probs + attention_probs_offset, v, current_tmp_data, nullptr);
+
+            // Transpose: out(B, S, N, H_v) -> out_tmp(B, N, S, H_v)
+            const int batch_index = static_cast<int>(i / num_heads);
+            const int head_index = static_cast<int>(i % num_heads);
+            T* src = current_tmp_data;
+            ptrdiff_t dest_offset =
+                (SafeInt<ptrdiff_t>(batch_index) * sequence_length * num_heads + head_index) * v_head_size;
+            T* dest = output + dest_offset;
+            for (int j = 0; j < sequence_length; j++) {
+              memcpy(dest, src, bytes_to_copy_trans);
+              src += v_head_size;
+              dest += v_hidden_size;
+            }
+          } else {
+            ptrdiff_t attention_probs_offset = SafeInt<ptrdiff_t>(sequence_length) * total_sequence_length * i;
+            ptrdiff_t dest_offset = SafeInt<ptrdiff_t>(sequence_length) * v_head_size * i;
+            T* dest = output + dest_offset;
+            math::MatMul<T>(sequence_length, v_head_size, total_sequence_length,
+                            attention_probs + attention_probs_offset, v, dest, nullptr);
+          }
+        }
+      });
+}
+
+template <typename T>
+Status AttentionBase<T>::ApplyAttention(OpKernelContext* context,
+                                        const T* Q,                             // Q data with shape BxNxSxH
+                                        const T* K,                             // K data with shape BxNxLxH
+                                        const T* V,                             // V value with size BxNxLxH_v
+                                        const Tensor* mask_index,               // mask index. nullptr if no mask or its size is B
+                                        const Tensor* past,                     // past state
```

Comment:
What is this input "past"? It seems nullptr in the only call in this implementation. What is it for?

Reply:
True, I forgot to remove, I used the code from the contribops as a baseline. I kept it because I thaught it would be needed for one the cases supported by the onnx ops. I forgot to remove.